lib/container/HoistContainerReferencesPlugin.js

```diff
@@ -0,0 +1,205 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Zackary Jackson @ScriptedAlchemy
+*/
+
+"use strict";
+
+const { AsyncDependenciesBlock, ExternalModule } = require("webpack");
+const ModuleFederationPlugin = require("./ModuleFederationPlugin");
+
+const PLUGIN_NAME = "HoistContainerReferences";
+
+/**
+ * This class is used to hoist container references in the code.
+ */
+class HoistContainerReferences {
+	/**
+	 * Apply the plugin to the compiler.
+	 * @param {import("webpack").Compiler} compiler The webpack compiler instance.
+	 */
+	apply(compiler) {
+		compiler.hooks.thisCompilation.tap(PLUGIN_NAME, compilation => {
+			const hooks = ModuleFederationPlugin.getCompilationHooks(compilation);
+			const containerEntryDependencies = new Set();
+			hooks.addContainerEntryModule.tap(PLUGIN_NAME, dep => {
+				containerEntryDependencies.add(dep);
+			});
+			hooks.addFederationRuntimeModule.tap(PLUGIN_NAME, dep => {
+				containerEntryDependencies.add(dep);
+			});
+
+			// Hook into the optimizeChunks phase
+			compilation.hooks.optimizeChunks.tap(
+				{
+					name: PLUGIN_NAME,
+					// advanced stage is where SplitChunksPlugin runs.
+					stage: 11 // advanced + 1
+				},
+				chunks => {
+					this.hoistModulesInChunks(compilation, containerEntryDependencies);
+				}
+			);
+		});
+	}
+
+	/**
+	 * Hoist modules in chunks.
+	 * @param {import("webpack").Compilation} compilation The webpack compilation instance.
+	 * @param {Set<import("webpack").Dependency>} containerEntryDependencies Set of container entry dependencies.
+	 */
+	hoistModulesInChunks(compilation, containerEntryDependencies) {
+		const { chunkGraph, moduleGraph } = compilation;
+		// when runtimeChunk: single is set - ContainerPlugin will create a "partial" chunk we can use to
+		// move modules into the runtime chunk
+		for (const dep of containerEntryDependencies) {
+			const containerEntryModule = moduleGraph.getModule(dep);
+			if (!containerEntryModule) continue;
+			const allReferencedModules = getAllReferencedModules(
+				compilation,
+				containerEntryModule,
+				"initial"
+			);
+
+			const allRemoteReferences = getAllReferencedModules(
```

Comment:
Get external modules types like remote modules and externals set in config - if user eager share  or consume module that needs external, it might not have loaded the chunk who had it in the entry so it needs hoisting too.

Reply:
Mostly an issue in single runtime chunk, but if user has library as well in their host code from federation v2, it will occur there